### PR TITLE
AMQ-6148 re-using LDAP context for authentication

### PR DESCRIPTION
**Pull request by Dm-Chebotarskyi**
### Description
- Re-using LDAP context for authentication.
- Refactoring some code in `LDAPLoginModule.java`

### Testing
Manual test authentication using LDAP `TBD`

### Issue
https://issues.apache.org/jira/browse/AMQ-6148

### CODE REVIEW

File: activemq-jaas/src/main/java/org/apache/activemq/jaas/LDAPLoginModule.java [hunk 1]
```diff
@@ -187,7 +183,10 @@ public boolean abort() throws LoginException {
         return true;
     }
 
-    protected void close(DirContext context) {
+    protected void closeContext() {
+        if (context == null) {
+            return;
+        }
         try {
             context.close();
```
Comment (epytka) [inline activemq-jaas/src/main/java/org/apache/activemq/jaas/LDAPLoginModule.java:191]: add `context = null;` after closing context

File: activemq-jaas/src/main/java/org/apache/activemq/jaas/LDAPLoginModule.java [hunk 2]
```diff
@@ -490,45 +482,47 @@ protected boolean bindUser(DirContext context, String dn, String password) throw
...
-        try {
-            Hashtable<String, String> env = new Hashtable<String, String>();
-            env.put(Context.INITIAL_CONTEXT_FACTORY, getLDAPPropertyValue(INITIAL_CONTEXT_FACTORY));
-            if (isLoginPropertySet(CONNECTION_USERNAME)) {
-                env.put(Context.SECURITY_PRINCIPAL, getLDAPPropertyValue(CONNECTION_USERNAME));
-            } else {
-                throw new NamingException("Empty username is not allowed");
-            }
+    protected void openContext() throws NamingException {
+        if (context == null) {
```
Comment (epytka) [inline activemq-jaas/src/main/java/org/apache/activemq/jaas/LDAPLoginModule.java]: to be consistent with `closeContext` we could use: 

```
if (context != null){
   return; 
}
```

**Review by epytka — CHANGES_REQUESTED** [review]
Please null context after closing it otherwise it wont be created again after close

File: activemq-jaas/src/main/java/org/apache/activemq/jaas/LDAPLoginModule.java (hunk 1 above)
Comment (Dm-Chebotarskyi) [reply activemq-jaas/src/main/java/org/apache/activemq/jaas/LDAPLoginModule.java:191]: fixed

File: activemq-jaas/src/main/java/org/apache/activemq/jaas/LDAPLoginModule.java (hunk 2 above)
Comment (Dm-Chebotarskyi) [reply activemq-jaas/src/main/java/org/apache/activemq/jaas/LDAPLoginModule.java]: fixed

File: activemq-jaas/src/main/java/org/apache/activemq/jaas/LDAPLoginModule.java [hunk 3]
```diff
@@ -320,20 +316,19 @@ protected boolean authenticate(String username, String password) throws LoginExc
...
-            if (context != null) {
-                close(context);
-            }
+            closeContext();
             FailedLoginException ex = new FailedLoginException("Error contacting LDAP");
             ex.initCause(e);
             throw ex;
         }
 
         if (context != null) {
```
Comment (epytka) [inline activemq-jaas/src/main/java/org/apache/activemq/jaas/LDAPLoginModule.java]: You don't need to check if context != null as in closeContext method it's already checked

**Comment by Charlie-chenchrl** [comment]
Any chance to add a test?

**Comment by Dm-Chebotarskyi** [comment]
> Any chance to add a test?

File: activemq-jaas/src/main/java/org/apache/activemq/jaas/LDAPLoginModule.java (hunk 3 above)
Comment (Dm-Chebotarskyi) [reply activemq-jaas/src/main/java/org/apache/activemq/jaas/LDAPLoginModule.java]: fixed

**Comment by Dm-Chebotarskyi** [comment]
> Any chance to add a test?

This functionality is already tested by existing test.

**Review by epytka — APPROVED** [review]

**Pull request merged**